modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2018-2020 ProfunKtor
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.profunktor.redis4cats
+
+/**
+  * An heterogeneous list, mainly used to operate on transactions.
+  *
+  * Highly inspired by Shapeless machinery but very much lightweight.
+  */
+object hlist {
+
+  type ::[H, T <: HList] = HCons[H, T]
+  type HNil              = HNil.type
+
+  sealed trait HList {
+    type Prepend[A] <: HList
+    def ::[A](a: A): Prepend[A]
+
+    def reverse: HList = {
+      def go(res: HList, ys: HList): HList =
+        ys match {
+          case HNil        => res
+          case HCons(h, t) => go(h :: res, t)
+        }
+      go(HNil, this)
+    }
+  }
+
+  case class HCons[H, Tail <: HList](head: H, tail: Tail) extends HList {
+    override type Prepend[A] = HCons[A, HCons[H, Tail]]
```

Comment:
Isn't `Prepend[A]` always going to be `HCons[A, this.type]`?

Reply:
Mmm no, I don't think that would work...

Reply:
It seems to compile:

```diff
diff --git a/modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala b/modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala
index 5588f88..82573e7 100644
--- a/modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala
+++ b/modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala
@@ -29,8 +29,7 @@ object hlist {
   type HNil              = HNil.type
 
   sealed trait HList {
-    type Prepend[A] <: HList
-    def ::[A](a: A): Prepend[A]
+    def ::[A](a: A): HCons[A, this.type] = HCons(a, this)
 
     def reverse: HList = {
       @tailrec
@@ -43,15 +42,9 @@ object hlist {
     }
   }
 
-  case class HCons[H, Tail <: HList](head: H, tail: Tail) extends HList {
-    override type Prepend[A] = HCons[A, HCons[H, Tail]]
-    override def ::[A](a: A): Prepend[A] = HCons(a, this)
-  }
+  case class HCons[H, Tail <: HList](head: H, tail: Tail) extends HList
 
-  case object HNil extends HList {
-    override type Prepend[A] = HCons[A, HNil]
-    override def ::[A](a: A): Prepend[A] = HCons(a, this)
-  }
+  case object HNil extends HList
 
   /**
     * It witnesses a relationship between two HLists.
```

Reply:
although it does seem to corrupt the examples. I'll give it another look.

Reply:
Okay, here's a version that compiles for both examples and the tests:

```diff
diff --git a/modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala b/modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala
index 5588f88..8770652 100644
--- a/modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala
+++ b/modules/core/src/main/scala/dev/profunktor/redis4cats/hlist.scala
@@ -29,8 +29,7 @@ object hlist {
   type HNil              = HNil.type
 
   sealed trait HList {
-    type Prepend[A] <: HList
-    def ::[A](a: A): Prepend[A]
+    def ::[A](a: A): HCons[A, this.type] = HCons(a, this)
 
     def reverse: HList = {
       @tailrec
@@ -43,15 +42,9 @@ object hlist {
     }
   }
 
-  case class HCons[H, Tail <: HList](head: H, tail: Tail) extends HList {
-    override type Prepend[A] = HCons[A, HCons[H, Tail]]
-    override def ::[A](a: A): Prepend[A] = HCons(a, this)
-  }
+  case class HCons[+H, +Tail <: HList](head: H, tail: Tail) extends HList
 
-  case object HNil extends HList {
-    override type Prepend[A] = HCons[A, HNil]
-    override def ::[A](a: A): Prepend[A] = HCons(a, this)
-  }
+  case object HNil extends HList
 
   /**
     * It witnesses a relationship between two HLists.

```

Reply:
Awesome! Just incorporated these changed :100: